satpy/readers/hdfeos_base.py

```diff
@@ -0,0 +1,240 @@
+from pyhdf.SD import SD
+
+from satpy import CHUNK_SIZE
+from satpy.readers.file_handlers import BaseFileHandler
+
+logger = logging.getLogger(__name__)
+
+
+class HDFEOSBaseFileReader(BaseFileHandler):
+    """Base file handler for HDF EOS data for both L1b and L2 products. """
+    def __init__(self, filename, filename_info, filetype_info):
+        BaseFileHandler.__init__(self, filename, filename_info, filetype_info)
+        try:
+            self.sd = SD(self.filename)
+        except HDF4Error as err:
+            error_message = "Could not load data from file {}: {}".format(self.filename, err)
+            raise ValueError(error_message)
+
+        # Read metadata
+        self.metadata = self.read_mda(self.sd.attributes()['CoreMetadata.0'])
+        self.metadata.update(self.read_mda(
+            self.sd.attributes()['StructMetadata.0'])
+        )
+        self.metadata.update(self.read_mda(
+            self.sd.attributes()['ArchiveMetadata.0'])
+        )
+
+    @staticmethod
+    def read_mda(attribute):
+        lines = attribute.split('\n')
+        mda = {}
+        current_dict = mda
+        path = []
+        prev_line = None
+        for line in lines:
+            if not line:
+                continue
+            if line == 'END':
+                break
+            if prev_line:
+                line = prev_line + line
+            key, val = line.split('=')
+            key = key.strip()
+            val = val.strip()
+            try:
+                val = eval(val)
+            except NameError:
+                pass
+            except SyntaxError:
+                prev_line = line
+                continue
+            prev_line = None
+            if key in ['GROUP', 'OBJECT']:
+                new_dict = {}
+                path.append(val)
+                current_dict[val] = new_dict
+                current_dict = new_dict
+            elif key in ['END_GROUP', 'END_OBJECT']:
+                if val != path[-1]:
+                    raise SyntaxError
+                path = path[:-1]
+                current_dict = mda
+                for item in path:
+                    current_dict = current_dict[item]
+            elif key in ['CLASS', 'NUM_VAL']:
+                pass
+            else:
+                current_dict[key] = val
+        return mda
+
+    @property
+    def start_time(self):
+        date = (self.metadata['INVENTORYMETADATA']['RANGEDATETIME']['RANGEBEGINNINGDATE']['VALUE'] + ' ' +
+                self.metadata['INVENTORYMETADATA']['RANGEDATETIME']['RANGEBEGINNINGTIME']['VALUE'])
+        return datetime.strptime(date, '%Y-%m-%d %H:%M:%S.%f')
+
+    @property
+    def end_time(self):
+        date = (self.metadata['INVENTORYMETADATA']['RANGEDATETIME']['RANGEENDINGDATE']['VALUE'] + ' ' +
+                self.metadata['INVENTORYMETADATA']['RANGEDATETIME']['RANGEENDINGTIME']['VALUE'])
+        return datetime.strptime(date, '%Y-%m-%d %H:%M:%S.%f')
+
+    @lru_cache(32)
```

Comment:
I'm curious about this: how much performance do you gain from it ? Using dask, I would have thought that each dataset would already be read only once.

Reply:
I wasn't aware of dask caching the dataset already. 
In https://github.com/pytroll/satpy/pull/540/files @BENR0 introduced a cache attribute to store already interpolated datasets. I tried to mimic this approach.

Reply:
@BENR0 , what was your intention ?

Reply:
Actually it was not me who introduced that. I took the  "HDFEOSFileReader" and "HDFEOSGeoReader" from the level 1 reader. The reason for the first one was that I needed to change the read_mda method due to the problem in the meta data of the mod35_l2 files. I can't remember why I also copied over the GeoReader right now. That's why there was so much room for refactorisation. I think in between my original PR and @LTMeyer working on the reader again the get_dataset and interpolation methods in the GeoReader method of the Level 1 reader changed quite a bit so it seems like the cache thing got introduced by my PR but originally they came from the l1b reader.

In short: if dask takes care of it, there would be no reason from my side to include the caching part.

For completeness sake I still would like to include the 250m dataset even though from my experience it is not very useable. But as a user I would expect a reader to be able to read all datesets contained in a file.

Otherwise the refactorisation is really nice work @LTMeyer.

Reply:
Thank you for this precision @BENR0, I will remove the cache part then. 

Regarding the 250m resolution dataset @djhoese has previously commented:
> I was talking about this with Kathy Strabala (Project Manager of IMAPP) and she brought up (from what she remembers) that a QA flag should also be used when generating the 250m version of the product to know if the mask was valid. Her recollection was that without checking the QA flag you can't be sure if the mask is "cloud" or invalid because the default pixel value is "cloud" or something like that.
Does this sound familiar? Are you handling a quality mask like this at all?

How should we integrate the QA flag to validate the cloud mask dataset? 
What behavior should the reader have in case of invalid dataset?

Reply:
You could load the QA flag separate and use xarray's `where` to mask any invalid pixels (which will set them to NaN). I could also see setting them to the "invalid" value of the mask to try to preserve the integer data type.

Reply:
I've looked into adding 250m resolution for the cloud mask. 
As a first step, I've added the QA among available dataset. It will then be used for 250m dataset validation.

I have concerns with the 250m cloud mask dataset, which requires more manipulation than lower resolution.
Indeed the information is stored in two different bytes, whose bits need to be stripped, validated by the QA mask, and then recombined as an array of right shape. In comparison, the 1km dataset doesn't need any validation nor reshaping.
Although these manipulations can be achieved by hard coding them in the reader, we would rather like to use parameters stored in the YAML file. What key value pair would make more sense? A `quality_assurance` boolean to use validation mask? A `reshape` list to reshape combined values of different bytes.

Reply:
I'm not sure I understand the reshape part of it. After masking bits you have two arrays of the same shape right?

Side note: Could you use `0b000011` like syntax to simplify the yaml? See https://yaml.org/type/int.html I can't tell from this page if the `!!int` part is needed or not.

Reply:
> I'm not sure I understand the reshape part of it. After masking bits you have two arrays of the same shape right (one for the cloud mask and one for the quality mask)?

The QA is 1km resolution, whereas the cloud mask, we are interested in, has a resolution of 250m. 

The information of the 250m resolution cloud mask is stored in the 8 bits of the fourth and fifth bytes of the `Cloud_Mask` dataset ([c.f. file description](https://modis-atmos.gsfc.nasa.gov/sites/default/files/ModAtmo/MYD35_L2.C6.CDL.fs)). For one "pixel" of the 1km resolution cloud mask, stand 16 (4x4) pixels for the 250m resolution dataset. Those pixels are organized as follows: 
- byte 4 bit 0 : Element(1,1)
- byte 4 bit 1 : Element(1,2)
- ...
- byte 5 bit 7 : Element(4,4)

Therefore, we need to reshape the original datasets to get a final array whose dimension are `4*SCAN_WIDTH` and `4*SCAN_LEN`. 

I have found the following working code in @BENR0's branch. The question is how to avoid hard coding it in the reader. 

```
# Cloud mask is stored in two bytes we need to recombine in one dataset
dataset_1 = dataset[4, :, :]
dataset_2 = dataset[5, :, :]

# We need to extract the value of each bit and return an array of relevant size

# Concatenate the two bytes
dataset_1 = np.uint16(dataset_1)
# Shift bits to the left
dataset_1 = np.left_shift(dataset_1, 8)  # dataset_1 << 8
dataset = da.bitwise_or(dataset_1, dataset_2).astype(np.uint16)

# Create 4x4 mask with numbers from 1 to 16 (count of bits) for extraction
bit_start = np.arange(16, dtype=np.uint16).reshape((4, 4))
# Tile the 4x4 array the size of the input array dimensions
bit_start = np.tile(bit_start, (dataset.shape[0], dataset.shape[1]))
# Replicate bits
bit_data = np.repeat(np.repeat(dataset, 4, axis=0), 4, axis=1)
# Decode the corresponding bits to get the 250m final dataset
dataset = bits_strip(bit_start, 1, bit_data)
```

> Side note: Could you use 0b000011 like syntax to simplify the yaml? See https://yaml.org/type/int.html I can't tell from this page if the !!int part is needed or not.

I'll have a look.

Reply:
For actually determining the value of the bit mask, I personally would do a bitwise and followed by a right shift, but that's after the bytes have been reconstructed. I'm betting I'll have to update whatever you end up doing to support cloud mask files from the IMAPP software where only 1km cloud mask is provide as a single byte field (to simplify the logic you are currently running in to).

The bit mask and shift should be YAML parameters in my opinion. I think having a value in the YAML to trigger the reshape is probably your best/easiest solution. To join the two bytes together...maybe it is just best to hardcode it with `if ds_id.name == "cloud_mask"` or some YAML parameter check to hide all of the magic under one `if` statement.

Reply:
> For actually determining the value of the bit mask, I personally would do a bitwise and followed by a right shift, but that's after the bytes have been reconstructed.

I believe this is what the code currently does. The `bitwise_or`and the `left_shit` are only used to reconstruct the bytes. Then the function `bits_strip`, which extracts the value of each bit,  does indeed a `bitwise_and` and `right_shift`operations. I should have added this code too for more clarity.

I will work on it based on your suggestion.